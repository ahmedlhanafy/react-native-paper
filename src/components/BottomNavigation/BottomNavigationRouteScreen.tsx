import React from 'react';
import { Animated, Platform, View, ViewProps } from 'react-native';

interface Props extends ViewProps {
  visibility?: 0 | 1 | Animated.AnimatedInterpolation;
  index: number;
}

class BottomNavigationRouteScreen extends React.Component<Props> {
  render(): JSX.Element {
    const { style, index, children, visibility, ...rest } = this.props;

    // Use display only on web, to fix animations during transitions between tabs.
    const display =
      Platform.OS === 'web' ? (visibility === 0 ? 'none' : 'flex') : undefined;

    return (
      <View
        testID={`RouteScreen: ${index}`}
        style={[style, { display }]}
        {...rest}
      >
        {children}
      </View>
    );
  }
}

export default Animated.createAnimatedComponent(BottomNavigationRouteScreen);
