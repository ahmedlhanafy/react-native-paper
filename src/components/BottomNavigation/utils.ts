import color from 'color';
import type { InternalTheme } from 'src/types';

import type { black, white } from '../../styles/themes/v2/colors';

type BaseConfig = {
  defaultColor: typeof black | typeof white;
  theme: InternalTheme;
};

export const getActiveTintColor = ({
  activeColor,
  defaultColor,
  theme,
}: {
  activeColor: string | undefined;
} & BaseConfig) => {
  return activeColor
    ? activeColor
    : theme.isV3
    ? theme.colors.onSecondaryContainer
    : defaultColor;
};

export const getInactiveTintColor = ({
  inactiveColor,
  defaultColor,
  theme,
}: {
  inactiveColor: string | undefined;
} & BaseConfig) => {
  return inactiveColor
    ? inactiveColor
    : theme.isV3
    ? theme.colors.onSurfaceVariant
    : color(defaultColor).alpha(0.5).rgb().string();
};

export const getLabelColor = ({
  tintColor,
  hasColor,
  focused,
  defaultColor,
  theme,
}: {
  tintColor: string;
  hasColor: boolean;
  focused: boolean;
} & BaseConfig) => {
  return hasColor
    ? tintColor
    : focused
    ? theme.colors.onSurface
    : theme.isV3
    ? theme.colors.onSurfaceVariant
    : defaultColor;
};
