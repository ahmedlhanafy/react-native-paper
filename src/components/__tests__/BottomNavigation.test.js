import * as React from 'react';
import renderer from 'react-test-renderer';
import { SafeAreaProvider } from 'react-native-safe-area-view';
import BottomNavigation from '../BottomNavigation.tsx';

jest.useFakeTimers();

jest.mock('Animated', () => {
  const ActualAnimated = jest.requireActual('Animated');

  return {
    ...ActualAnimated,
    timing: (value, config) => ({
      start: callback => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
    }),
  };
});

const icons = ['magnify', 'camera', 'inbox', 'heart', 'shopping-music'];

const createState = (index, length) => ({
  index,
  routes: Array.from({ length }, (_, i) => ({
    key: `key-${i}`,
    icon: icons[i],
    title: `Route: ${i}`,
  })),
});

it('renders shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting
          navigationState={createState(0, 5)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders non-shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting={false}
          navigationState={createState(0, 3)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders custom icon and label in shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting
          navigationState={createState(0, 5)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
          renderIcon={({ route, color }) => (
            <icon color={color}>{route.icon}</icon>
          )}
          renderLabel={({ route, color }) => (
            <text color={color}>{route.label}</text>
          )}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders custom icon and label in non-shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting={false}
          navigationState={createState(0, 3)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
          renderIcon={({ route, color }) => (
            <icon color={color}>{route.icon}</icon>
          )}
          renderLabel={({ route, color }) => (
            <text color={color}>{route.label}</text>
          )}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders custom icon and label with custom colors in shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting
          navigationState={createState(0, 3)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
          activeColor="#FBF7DB"
          inactiveColor="#853D4B"
          barStyle={{ backgroundColor: '#E96A82' }}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders custom icon and label with custom colors in non-shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting={false}
          navigationState={createState(0, 3)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
          activeColor="#FBF7DB"
          inactiveColor="#853D4B"
          barStyle={{ backgroundColor: '#E96A82' }}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('hides labels in shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting
          labeled={false}
          navigationState={createState(0, 3)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('hides labels in non-shifting bottom navigation', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider>
        <BottomNavigation
          shifting={false}
          labeled={false}
          navigationState={createState(0, 3)}
          onIndexChange={jest.fn()}
          renderScene={({ route }) => route.title}
        />
      </SafeAreaProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
