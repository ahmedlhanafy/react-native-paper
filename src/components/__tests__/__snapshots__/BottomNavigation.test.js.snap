// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hides labels in non-shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`hides labels in shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders custom icon and label in non-shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders custom icon and label in shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders custom icon and label with custom colors in non-shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders custom icon and label with custom colors in shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders non-shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders shifting bottom navigation 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
